[gym] Add option to pass in specific toolchain

nafu opened the pull request:
- This adds features for https://github.com/fastlane/fastlane/issues/1828,  https://github.com/fastlane/fastlane/issues/1079, https://github.com/fastlane/fastlane/issues/5863
  - Add toolchain option to BuildCommandGenerator and PackageCommandGeneratorXcode7

You can now use toolchain option like this:

``` ruby
gym(
  workspace: WORKSPACE,
  scheme: SCHEME,
  toolchain: "com.apple.dt.toolchain.Swift_2_3",
  configuration: "AdHoc",
  use_legacy_build_api: false,
  export_method: "ad-hoc"
)
```

nafu committed:
[gym] Add toolchain option to BuildCommandGenerator

nafu committed:
[gym] Add toolchain option to PackageCommandGeneratorXcode7

bfcrampton commented:
@nafu Looks like you also need to add it to `package_command_generator_legacy.rb` [like this](https://github.com/fastlane/fastlane/compare/master...bfcrampton:gym-toolchain-option#diff-2f3bd8439ae935dc4650d9242f2fad03R18) to work if `use_legacy_build_api: true` is set.

nafu commented:
@bfcrampton 

Thanks for your comment.
I think legacy build api doesn't accept toolchain option.

```bash
$ xcrun /tmp/PackageApplication4Gym --help



warning: PackageApplication is deprecated, use `xcodebuild -exportArchive` instead.


Usage:
    PackageApplication [-s signature] application [-o output_directory]
    [-verbose] [-plugin plugin] || -man || -help

    Options:

        -s <signature>  certificate name to resign application before packaging
        -o              specify output filename
        -plugin         specify an optional plugin
        -help           brief help message
        -man            full documentation
        -v[erbose]      provide details during operation
```

I'll add codes to print messages in [print_legacy_information ](https://github.com/fastlane/fastlane/blob/0c37f2313566e1a701cd93a20e44d196f71c98b7/gym/lib/gym/generators/package_command_generator_legacy.rb#L83-L99).

nafu committed:
[gym] Add warnings for toolchain option

Add warnings for toolchain option when using
Xcode's 7 (and above) build options with legacy build api.

KrauseFx commented on gym/lib/gym/options.rb:
```diff
@@ -158,6 +158,10 @@ def self.plain_options
                                      optional: true),
+        FastlaneCore::ConfigItem.new(key: :toolchain,
+                                     env_name: "GYM_TOOLCHAIN",
+                                     description: "The toolchain that should be used for building the application",
```
Could you provide some more information of what a `toolchain` is?

KrauseFx replied:
Is there a good way to validate the input, or to show some of the available values?

nafu replied:
@KrauseFx 

> Could you provide some more information of what a toolchain is?

From [Swift.org](https://swift.org/download/#using-downloads)

> An Xcode toolchain (.xctoolchain) includes a copy of the compiler, lldb, and other related tools needed to provide a cohesive development experience for working in a specific version of Swift.

Should I add this in description of gym options?

e.g.
```ruby
FastlaneCore::ConfigItem.new(key: :toolchain,
                             env_name: "GYM_TOOLCHAIN",
                             description: "The toolchain that should be used for building the application. An Xcode toolchain (.xctoolchain) includes a copy of the compiler, lldb, and other related tools needed to provide a cohesive development experience for working in a specific version of Swift",
                             optional: true),
```

> Is there a good way to validate the input, or to show some of the available values?

I have no idea. :cry:
Although we can get available value from `/Library/Developer/Toolchains/` and `/Applications/Xcode.app/Contents/Developer/Toolchains` by searching plist files, but it's not so simple and I think it's tough to validate it.

KrauseFx replied:
Alright that's fair

KrauseFx commented:
Hey @nafu, thanks for your pull request, this looks great. I added one comment before it's ready 👍

KrauseFx commented on gym/spec/build_command_generator_spec.rb:
```diff
@@ -22,7 +22,7 @@
         "#{k.to_s.shellescape}=#{v.shellescape}"
       end.join ' '
-      options = { project: "./examples/standard/Example.xcodeproj", sdk: "9.0", xcargs: xcargs, scheme: 'Example' }
+      options = { project: "./examples/standard/Example.xcodeproj", sdk: "9.0", toolchain: "com.apple.dt.toolchain.Swift_2_3", xcargs: xcargs, scheme: 'Example' }
```
Does the toolchain always start with `com.apple.dt.toolchain`? If so, we could automatically prefix user values 👍

nafu replied:
@KrauseFx 

> Does the toolchain always start with com.apple.dt.toolchain?

I don't think so.
Toolchain itself can be anything.

nafu replied:
Examples about toolchains.
It can be also org.swift.xxxx.

| With Xcode 8 Toolchain | env |
| --- | --- |
| <img width="360" src="https://cloud.githubusercontent.com/assets/932290/18035150/52226884-6d8a-11e6-90a2-ab7d46940ec0.png"> | ![screen shot 2016-08-29 at 1 43 53 am](https://cloud.githubusercontent.com/assets/932290/18035151/54a60aa2-6d8a-11e6-978d-37e79125e5ff.png) |

| With Swift 3.0 Preview 6 Snapshot | env |
| --- | --- |
| <img width="360" src="https://cloud.githubusercontent.com/assets/932290/18035171/c141d07e-6d8a-11e6-8928-3fa5af10561c.png"> | ![screen shot 2016-08-29 at 1 43 08 am](https://cloud.githubusercontent.com/assets/932290/18035172/c253759e-6d8a-11e6-9d99-961176968034.png) |

KrauseFx commented:
Hey @nafu, thanks for clarifying, this looks great to me. Could you only update the description of the new value, with an example value of what you can set?

Thanks 👍

nafu committed:
[gym] Add example values in description of toolchain option

nafu commented:
@KrauseFx 

I added example values.
https://github.com/fastlane/fastlane/pull/5864/commits/4c5da69e74e0a71b205516c3c2fd0702e4f213a0

Is this what you meant above? Let me know if I misunderstood anything.

KrauseFx commented on gym/lib/gym/generators/package_command_generator_xcode7.rb:
```diff
@@ -28,6 +28,10 @@ def options
         options << "-exportPath '#{temporary_output_path}'"
 
+        if Gym.config[:toolchain]
+          options << "-toolchain '#{Gym.config[:toolchain]}'"
```
So this value gets passed to both `xcodebuild` commands, for building and for archiving?

nafu replied:
Yes

KrauseFx commented:
Great, thanks for updating. I added one last question before we can merge 👍

KrauseFx commented:
Great work, thank you @nafu, this will be in the next release 🚀

nafu commented:
@KrauseFx 

Thank you, I appreciate it.

gbejarano01 commented:
Hi @nafu @KrauseFx I'm currently struggling with this too. Do you know if this has been added to version `1.102.0`?

Thanks!

KrauseFx commented:
@gbejarano01 There was no `gym` release yet https://rubygems.org/gems/gym

gbejarano01 commented:
Thanks @KrauseFx, that's what I thought, do you know when this can be included in a new version?

KrauseFx commented:
@gbejarano01 we're working on it 🚀

gbejarano01 commented:
Nice! Thanks @KrauseFx!
